### PR TITLE
W2HW3PR4 inference with parallelization

### PR DESCRIPTION
**Pull request by yevhen-k**
Hi @truskovskiyk 
Here is some experiments with data preprocessing parallelization. I've decided to use quite low-level approach with `multiprocessing.Process` and `multiprocessing.Queue`. It appeared, that for bert model with tokenizer a bottleneck is the inference stage.

### CODE REVIEW

File: week-2-storage/README.md [hunk 1]
```diff
@@ -209,4 +209,34 @@ kubectl port-forward service/mytenant-console 9090:9090 -n minio-tenant-1
...
+```
+Number of cpus                      : 8
+Dataset len                         : 2834
+
+[Single Process] Preprocessing      : 00:00:05
+[Single Process] Inference          : 00:01:56
+[Single Process] Total              : 00:02:02
+
+[Multiple Processes] Preprocessing  : 00:00:01
+[Multiple Processes] Inference      : 00:01:52
```
Comment (truskovskiyk) [inline week-2-storage/README.md:236]: looks like `Multiple Processes` is slower.

Comment (truskovskiyk) [reply week-2-storage/README.md:236]: any reasons why?

File: week-2-storage/test_parallelization.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+    def forward(self, x: Dict[str, torch.Tensor]) -> torch.Tensor:
+        features = self.base_model(**x)
+        output = torch.mean(features.last_hidden_state, dim=-2)
+        output = self.linear(output)
+        return output
+
+
+def init_model() -> Tuple[torch.nn.Module, torch.nn.Module, torch.device]:
+    if not os.path.exists("./weights"):
```
Comment (truskovskiyk) [inline week-2-storage/test_parallelization.py]: try to use https://docs.python.org/3/library/pathlib.html, much better than os.path

File: week-2-storage/test_parallelization.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        os.mkdir("./weights")
+    if not os.path.exists("./weights/bert-base-uncased-regression-weights.pth"):
+        gdown.download(
+            fuzzy=True, 
+            url="https://drive.google.com/file/d/1M5SFB5cYS7Q3oLpkEQVGERXtquwPxury/view?usp=sharing",
```
Comment (truskovskiyk) [inline week-2-storage/test_parallelization.py]: would be nice to move this into constants

File: week-2-storage/test_parallelization.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        start = i*dataset_per_cpu
+        end = (i+1)*dataset_per_cpu
+        if (dataset_size - end) >= dataset_per_cpu:
+            p = multiprocessing.Process(target=text_preprocessing, args=(df[start : end], queue))
+            jobs.append(p)
+            queues.append(queue)
+        else:
+            p = multiprocessing.Process(target=text_preprocessing, args=(df[start : ], queue))
+            jobs.append(p)
+            queues.append(queue)
```
Comment (truskovskiyk) [inline week-2-storage/test_parallelization.py:217]: curious why you did not get any speedup, what if we increase dataset size? 


what is reasoning to Queue btw? should be process pool easy to work with

**Review by truskovskiyk — COMMENTED** [review]
@yevhen-k looks great! please address comments!

File: week-2-storage/test_parallelization.py (hunk 4 above)
Comment (yevhen-k) [reply week-2-storage/test_parallelization.py:217]: Data processing has 3 steps:
- preporcessing (parallelization dropped down execution time from 5 to 1 second)
- tokenization (managed by pytorch and I didn't manage to parallelize it even with torch multiprocessing)
- inference (actually call the model with tokens to predict scores, parallelization is done by batching mechanism)

So, in this particular setup I can parallelize only preprocessing stage.

A few words about Queue. I know that it is better to use more high level abstractions like ProcessExecutor, but I just used to use Processes with Queues.

File: week-2-storage/README.md (hunk 1 above)
Comment (yevhen-k) [reply week-2-storage/README.md:236]: Data processing has 3 steps:
- preporcessing (parallelization dropped down execution time from 5 to 1 second)
- tokenization (not paralleled, see comments blow)
- inference (not paralleled, see comments blow)

Because the experiment was run only once, the phenomenon of slower multiprocessing is just a fluctuation of observation. So, we should consider preprocessing parallelization stage only.

File: week-2-storage/test_parallelization.py (hunk 2 above)
Comment (yevhen-k) [reply week-2-storage/test_parallelization.py]: Code updated

File: week-2-storage/test_parallelization.py (hunk 3 above)
Comment (yevhen-k) [reply week-2-storage/test_parallelization.py]: Code updated

File: week-2-storage/README.md (hunk 1 above)
Comment (truskovskiyk) [reply week-2-storage/README.md:236]: Got it, thanks for the exmplanation: 

5 seconds to 1-second looks like a marginal improvement, could we try to run this with a bigger dataset?

File: week-2-storage/test_parallelization.py (hunk 4 above)
Comment (truskovskiyk) [reply week-2-storage/test_parallelization.py:217]: got it, thanks for emplantion

File: week-2-storage/README.md (hunk 1 above)
Comment (yevhen-k) [reply week-2-storage/README.md:236]: Here are the results on a bigger data set for preprocessing stage only.

Dataset len: 14,170
Number of cpu : 8
[Single Process] Preprocessing: 00:00:20
[Multiple Processes] Preprocessing: 00:00:05
___

Dataset len: 141,700
Number of cpu : 8
[Single Process] Preprocessing: 00:03:15
[Multiple Processes] Preprocessing: 00:01:03

___

So, python multiprocessing sucks as usually. Having 8 processes without synchronization overhead I've managed to get only x3 increase by speed.

**Review by truskovskiyk — APPROVED** [review]
@yevhen-k thanks, approved!
